chain/checkpointing/util.go

```diff
@@ -0,0 +1,288 @@
+package checkpointing
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/Zondax/multi-party-sig/pkg/math/curve"
+	"github.com/btcsuite/btcutil/bech32"
+	"github.com/cronokirby/safenum"
+)
+
+type VerificationShare struct {
+	Share string
+}
+
+type TaprootConfigTOML struct {
+	Thershold          int
+	PrivateShare       string
+	PublicKey          string
+	VerificationShares map[string]VerificationShare
+}
+
+func TaggedHash(tag string, datas ...[]byte) []byte {
+	tagSum := sha256.Sum256([]byte(tag))
+
+	h := sha256.New()
+	h.Write(tagSum[:])
+	h.Write(tagSum[:])
+	for _, data := range datas {
+		h.Write(data)
+	}
+	return h.Sum(nil)
+}
+
+func Sha256(data []byte) []byte {
+	h := sha256.New()
+	h.Write(data[:])
+	return h.Sum(nil)
+}
+
+func TaprootSignatureHash(tx []byte, utxo []byte, hash_type byte) ([]byte, error) {
```

Comment:
@sa8 may I ask for you for some help reviewing this part? I am not super familiar with our design and I may be missing some details. Happy to do it sync together if it works best for you (that way I can learn a bit more about how the protocol works :) ).

Reply:
@sa8 Here the BIP 341 detailing how a message should be serialized before being signed (https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message). The scheme for Taproot has changed and is a bit more complicated.

You can also look at this python script used to test against the golang function (https://github.com/Zondax/fil-taproot/blob/main/scripts/main.py).

This step is being done right before signing and it return a hash. The hash is the message then signed by the participants.